File: src/helpers/calcStandings4.js
//Functional version of the code, where the assignment with the += operator is
//replaced with object spread and ternary conditional operators.
/*
const calcStandings4 = (matchResults) => {
    const standings = {};

    matchResults.forEach(match => {
        const { homeTeam, awayTeam, homeScore, awayScore, homeCrest, awayCrest } = match;
        
        standings[homeTeam] = {
            ...standings[homeTeam],
            crest: homeCrest,
            gp: (standings[homeTeam]?.gp || 0) + 1,
            goalsScored: (standings[homeTeam]?.goalsScored || 0) + homeScore,
            goalsConceded: (standings[homeTeam]?.goalsConceded || 0) + awayScore,
            points: (standings[homeTeam]?.points || 0) + (homeScore > awayScore ? 3 : homeScore === awayScore ? 1 : 0),
        };

        standings[awayTeam] = {
            ...standings[awayTeam],
            crest: awayCrest,
            gp: (standings[awayTeam]?.gp || 0) + 1,
            goalsScored: (standings[awayTeam]?.goalsScored || 0) + awayScore,
            goalsConceded: (standings[awayTeam]?.goalsConceded || 0) + homeScore,
            points: (standings[awayTeam]?.points || 0) + (awayScore > homeScore ? 3 : awayScore === homeScore ? 1 : 0),
        };
    });

    const sortedStandings = Object.keys(standings)
        .map(team => ({ team, ...standings[team] }))
        .sort((a, b) => b.points - a.points || (b.goalsScored - b.goalsConceded) - (a.goalsScored - a.goalsConceded));

    return sortedStandings;
};

export default calcStandings4;
*/

import _ from 'lodash';

const calcStandings4 = (matchResults) => {
  const standings = {};

  matchResults.forEach((match) => {
    const { homeTeam, awayTeam, homeScore, awayScore, homeCrest, awayCrest } =
      match;

    // determine the result of the match for each team
    // let homeResult, awayResult;
    // if (homeScore > awayScore) {
    //     homeResult = 'w';
    //     awayResult = 'l';
    // } else if (homeScore < awayScore) {
    //     homeResult = 'l';
    //     awayResult = 'w';
    // } else {
    //     homeResult = 'd';
    //     awayResult = 'd';
    // }

    // determine the result of the match for each team (using lodash)
    const homeResult = _.cond([
      [_.gt, _.constant("w")], // if homeScore > awayScore, return 'w'
      [_.lt, _.constant("l")], // if homeScore < awayScore, return 'l'
      [_.stubTrue, _.constant("d")], // otherwise, return 'd'
    ])(homeScore, awayScore); // apply the function to homeScore and awayScore

    const awayResult = _.cond([
      [_.gt, _.constant("l")], // if homeScore > awayScore, return 'l'
      [_.lt, _.constant("w")], // if homeScore < awayScore, return 'w'
      [_.stubTrue, _.constant("d")], // otherwise, return 'd'
    ])(homeScore, awayScore); // apply the function to homeScore and awayScore

    standings[homeTeam] = {
      ...standings[homeTeam],
      crest: homeCrest,
      gp: (standings[homeTeam]?.gp || 0) + 1,
      goalsScored: (standings[homeTeam]?.goalsScored || 0) + homeScore,
      goalsConceded: (standings[homeTeam]?.goalsConceded || 0) + awayScore,
      points:
        (standings[homeTeam]?.points || 0) +
        (homeScore > awayScore ? 3 : homeScore === awayScore ? 1 : 0),
      // append the result to the form attribute or create a new one if it does not exist
      form: (standings[homeTeam]?.form || "") + homeResult,
    };

    standings[awayTeam] = {
      ...standings[awayTeam],
      crest: awayCrest,
      gp: (standings[awayTeam]?.gp || 0) + 1,
      goalsScored: (standings[awayTeam]?.goalsScored || 0) + awayScore,
      goalsConceded: (standings[awayTeam]?.goalsConceded || 0) + homeScore,
      points:
        (standings[awayTeam]?.points || 0) +
        (awayScore > homeScore ? 3 : awayScore === homeScore ? 1 : 0),
      // append the result to the form attribute or create a new one if it does not exist
      form: (standings[awayTeam]?.form || "") + awayResult,
    };
  });

  const sortedStandings = Object.keys(standings)
    .map((team) => ({ team, ...standings[team] }))
    .sort(
      (a, b) =>
        b.points - a.points ||
        b.goalsScored - b.goalsConceded - (a.goalsScored - a.goalsConceded)
    );

  return sortedStandings;
};

export default calcStandings4;

File: src/helpers/calcStandings4.rs
fn calc_standings4(match_results: &Vec<HashMap<&str, &str>>) -> Vec<HashMap<&str, &str>> {
    let mut standings: HashMap<&str, HashMap<&str, i32>> = HashMap::new();

    for match_result in match_results {
        let home_team = match_result.get("homeTeam").unwrap();
        let away_team = match_result.get("awayTeam").unwrap();
        let home_score = match_result
            .get("homeScore")
            .unwrap()
            .parse::<i32>()
            .unwrap();
        let away_score = match_result
            .get("awayScore")
            .unwrap()
            .parse::<i32>()
            .unwrap();
        let home_crest = match_result.get("homeCrest").unwrap();
        let away_crest = match_result.get("awayCrest").unwrap();

        *standings.entry(home_team).or_insert(HashMap::new()) = {
            let mut home_team_stats = standings.entry(home_team).or_insert(HashMap::new());
            home_team_stats.insert("crest", home_crest.parse::<i32>().unwrap());
            home_team_stats.insert("gp", home_team_stats.get("gp").unwrap_or(&0) + 1);
            home_team_stats.insert(
                "goalsScored",
                home_team_stats.get("goalsScored").unwrap_or(&0) + home_score,
            );
            home_team_stats.insert(
                "goalsConceded",
                home_team_stats.get("goalsConceded").unwrap_or(&0) + away_score,
            );
            home_team_stats.insert(
                "points",
                home_team_stats.get("points").unwrap_or(&0)
                    + if home_score > away_score {
                        3
                    } else if home_score == away_score {
                        1
                    } else {
                        0
                    },
            );
            home_team_stats.clone()
        };

        *standings.entry(away_team).or_insert(HashMap::new()) = {
            let mut away_team_stats = standings.entry(away_team).or_insert(HashMap::new());
            away_team_stats.insert("crest", away_crest.parse::<i32>().unwrap());
            away_team_stats.insert("gp", away_team_stats.get("gp").unwrap_or(&0) + 1);
            away_team_stats.insert(
                "goalsScored",
                away_team_stats.get("goalsScored").unwrap_or(&0) + away_score,
            );
            away_team_stats.insert(
                "goalsConceded",
                away_team_stats.get("goalsConceded").unwrap_or(&0) + home_score,
            );
            away_team_stats.insert(
                "points",
                away_team_stats.get("points").unwrap_or(&0)
                    + if away_score > home_score {
                        3
                    } else if away_score == home_score {
                        1
                    } else {
                        0
                    },
            );
            away_team_stats.clone()
        };
    }

    let mut sorted_standings: Vec<HashMap<&str, &str>> = standings
        .iter()
        .map(|(team, stats)| {
            stats
                .iter()
                .map(|(k, v)| (*k, v.to_string().as_str()))
                .chain(std::iter::once(("team", team)))
                .collect()
        })
        .collect();

    sorted_standings.sort_by(|a, b| {
        b.get("points")
            .unwrap()
            .parse::<i32>()
            .unwrap()
            .cmp(&a.get("points").unwrap().parse::<i32>().unwrap())
            .then_with(|| {
                (b.get("goalsScored").unwrap().parse::<i32>().unwrap()
                    - b.get("goalsConceded").unwrap().parse::<i32>().unwrap())
                .cmp(
                    &(a.get("goalsScored").unwrap().parse::<i32>().unwrap()
                        - a.get("goalsConceded").unwrap().parse::<i32>().unwrap()),
                )
            })
    });

    sorted_standings
}
